File: gui/app/ConfigLoader.py
from __future__ import annotations
import json
from dataclasses import dataclass
from pathlib import Path
from typing import List


@dataclass
class DeviceConfig:
    uid:  str
    type: str
    port: int


class ConfigLoader:
    def __init__(self, config_path: str) -> None:
        self._path    = Path(config_path)
        self.devices: List[DeviceConfig] = []

    def load(self) -> None:
        if not self._path.exists():
            raise FileNotFoundError(f"Config not found: {self._path}")
        with open(self._path) as f:
            data = json.load(f)
        if "devices" not in data or not isinstance(data["devices"], list):
            raise ValueError("Config missing 'devices' array")
        self.devices = [
            DeviceConfig(uid=d["uid"], type=d["type"], port=int(d["port"]))
            for d in data["devices"]
        ]

    def get_devices(self) -> List[DeviceConfig]:
        return self.devices

File: monitor/src/ConfigLoader.cpp
#include "ConfigLoader.hpp"
#include <fstream>
#include <stdexcept>
#include <nlohmann/json.hpp>

ConfigLoader::ConfigLoader(const std::string& config_path)
    : config_path_(config_path) {}

void ConfigLoader::load() {
    std::ifstream f(config_path_);
    if (!f.is_open())
        throw std::runtime_error("Cannot open config: " + config_path_);

    nlohmann::json j;
    try {
        f >> j;
    } catch (const nlohmann::json::exception& e) {
        throw std::runtime_error("Config JSON parse error: " + std::string(e.what()));
    }

    if (!j.contains("devices") || !j["devices"].is_array())
        throw std::runtime_error("Config missing 'devices' array");

    devices_.clear();
    for (const auto& d : j["devices"]) {
        if (!d.contains("uid") || !d.contains("type") || !d.contains("port"))
            throw std::runtime_error("Device entry missing uid/type/port fields");

        DeviceConfig cfg;
        cfg.uid  = d["uid"].get<std::string>();
        cfg.type = d["type"].get<std::string>();
        cfg.port = static_cast<uint16_t>(d["port"].get<int>());
        devices_.push_back(std::move(cfg));
    }
}
